File: DataAlgo_forPy/Search_Sort/binary_search.rb
def binary_search(array, target)
  first = 0
  last = array.size - 1
  found = false

  while first <= last && !found
    mid = (first + last) / 2

    if target == array[mid]
      found = true
    elsif target > array[mid]
      first = mid + 1
    else
      last = mid - 1
    end

  end

  found
end

def recursive_bsearch(array,target)
  return false if array.empty?

  mid = array.size / 2
  element = array[mid]
  left = array[0...mid]
  right = array[(mid+1)...-1]

  case target <=> element
  when 0
    true
  when 1
    recursive_bsearch(right, target)
  when -1
    recursive_bsearch(left, target)
  end
end



p binary_search([1,2,3,4,5,6,7], 5)
p recursive_bsearch([1,2,3,4,5,6,7], 5)

File: DataAlgo_forPy/Search_Sort/binary_search.js
const binary_search = (array, target) => {
    let first = 0
    let last = array.length - 1
    let found = false

    while (first <= last && !found) {
        let mid = Math.trunc((first + last) / 2)

        if (array[mid] === target) {
            found = true
        } else if (target < array[mid]) {
            last = mid - 1
        } else {
            first = mid + 1
        }

    }

    return found
}

const recursive_bsearch = (array, target) => {
  if (array.length === 0) { return false }

  const mid = Math.trunc(array.length / 2)
  const element = array[mid]
  const left = array.slice(0, mid)
  const right = array.slice(mid + 1)

  if (element === target) {
    return true
  } else if (target < element) {
    return recursive_bsearch(left, target)
  } else {
    return recursive_bsearch(right, target)
  }

}

console.log(binary_search([1,2,3,4,5,6,7], 5))
console.log(recursive_bsearch([1,2,3,4,5,6,7], 5))
